marconi-core/src/Marconi/Core/Transformer/WithCatchup.hs

```diff
@@ -191,48 +196,58 @@ instance
   (MonadIO m, IsIndex m event indexer, Ord (Point event))
   => IsIndex m event (WithCatchup indexer)
   where
-  index timedEvent@(Timed p e) indexer =
-    let bufferIsFull ix = (ix ^. catchupBufferLength) >= (ix ^. catchupBatchSize)
-        pushEvent ix =
-          ix
-            & catchupBuffer %~ (timedEvent :)
-            & catchupBufferLength +~ 1
-        hasCaughtUp = case e of
-          Nothing -> False
-          Just e' -> (indexer ^. catchupDistance) p e' < indexer ^. catchupBypassDistance
-        sendBatch ix = do
-          ix' <- indexAllDescendingVia caughtUpIndexer (ix ^. catchupBuffer) ix
-          ix'' <- case ix' ^. catchupLastStable of
-            Nothing -> pure ix'
-            Just lastStable -> setLastStablePointVia caughtUpIndexer lastStable ix'
-          pure $ resetBuffer ix''
-     in if hasCaughtUp
-          then do
-            maybe (pure ()) (\f -> liftIO $ f Synced) $ indexer ^. catchupEventHook
-            indexer' <-
-              if null (indexer ^. catchupBuffer)
-                then pure indexer
-                else sendBatch indexer
-            indexVia caughtUpIndexer timedEvent indexer'
-          else do
-            let indexer' = pushEvent indexer
-            if bufferIsFull indexer'
-              then sendBatch indexer'
-              else pure indexer'
+  index timedEvent@(Timed p e) indexer = case indexer ^. config of
+    Done -> indexVia caughtUpIndexer timedEvent indexer
+    Ongoing cfg -> do
+      let batchSize ix = ix ^? catchupBatchSize
+          bufferLength ix = ix ^? catchupBufferLength
+          bufferIsFull ix = fromMaybe False $ do
+            l <- bufferLength ix
+            s <- batchSize ix
+            pure $ l >= s
+          bypassDistance = cfg ^. contextCatchupConfig . configCatchupBypassDistance
+          pushEvent ix =
+            ix
+              & catchupBuffer %~ (timedEvent :)
+              & catchupBufferLength +~ 1
+          hasCaughtUp = case e of
+            Nothing -> False
+            Just e' -> (cfg ^. contextDistanceComputation) p e' < bypassDistance
+          sendBatch ix = do
+            let catchupBuffer' = ix ^. catchupBuffer
+            let lastStable = ix ^? catchupLastStable
+            ix' <- indexAllDescendingVia caughtUpIndexer catchupBuffer' ix
+            ix'' <- case lastStable of
+              Nothing -> pure ix'
+              Just lastStable' -> setLastStablePointVia caughtUpIndexer lastStable' ix'
+            pure $ resetBuffer ix''
+          triggerHook = case cfg ^. contextCatchupEventHook of
+            Nothing -> pure ()
+            Just hook -> liftIO $ hook Synced
+      if hasCaughtUp
+        then do
+          triggerHook
+          indexer' <- sendBatch indexer
```

Comment:
`sendBatch` in some cases calls `setLastStablePointVia`. In the previous version of this section, the last stable point was not updated when the catchup buffer was empty: The original indexer was returned.

Was that intended? From what I see I think it is a small but inoffensive difference: The last stable point might be updated more frequently.

The remaining stuff in `sendBatch` is a no-op when the buffer is empty, I think.

Reply:
It may call it one time more. `hasCaughtUp` should be true only one time during synchronisation: when we reach the expected distance to the tip.